### PR TITLE
Added a way for easy mal and anilist tracker

### PR DESCRIPTION
**Pull request by KillerDogeEmpire**
All credit gos to Hexated for helping me

### CODE REVIEW

**Comment by C10udburst** [comment]
[CodeFactor](https://www.codefactor.io/repository/github/recloudstream/cloudstream) found multiple issues last seen at c7c1f12dade36a515fb7a09ececc3f1d10a7b0d1:

#### Parentheses in (media.type.equals(type, true) &amp;&amp; media.releaseDate == year) are unnecessary and can be replaced with: media.type.equals(type, true) &amp;&amp; media.releaseDate == year
- [ ]  [app\src\main\java\com\lagradost\cloudstream3\MainAPI.kt:172](https://github.com/recloudstream/cloudstream/blob/c7c1f12dade36a515fb7a09ececc3f1d10a7b0d1/app/src/main/java/com/lagradost/cloudstream3/MainAPI.kt#L172)

#### Parentheses in (media.title?.english.equals(title, true) || media.title?.romaji.equals(                     title,                     true                 )) are unnecessary and can be replaced with: media.title?.english.equals(title, true) || media.title?.romaji.equals(                     title,                     true                 )
- [ ]  [app\src\main\java\com\lagradost\cloudstream3\MainAPI.kt:169](https://github.com/recloudstream/cloudstream/blob/c7c1f12dade36a515fb7a09ececc3f1d10a7b0d1/app/src/main/java/com/lagradost/cloudstream3/MainAPI.kt#L169)

File: app/src/main/java/com/lagradost/cloudstream3/MainAPI.kt [hunk 1]
```diff
@@ -163,6 +163,38 @@ object APIHolder {
...
+     * Both titles are attempted to be matched with both Romaji and English title.
+     * Uses the consumet api.
+     *
+     * @param mainTitle Title to search by and match
+     * @param secondaryTitle Optional extra title if you have multiple titles and want extra guarantee to match.
+     * @param types Optional parameter to narrow down the scope to Movies, TV, etc. See TrackerType.getTypes()
+     * @param year Optional parameter to only get anime with a specific year
+     **/
+    suspend fun getTracker(
+        mainTitle: String,
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/MainAPI.kt:177]: do a List<String> instead here, then smth like matchingTitles = media.title?.let { title -> titles.any { userTitle -> title.isMatchingTitles(userTitle) } } :? true

this is to match any number of titles

Comment (LagradOst) [reply app/src/main/java/com/lagradost/cloudstream3/MainAPI.kt:177]: and uses the first title for search

File: app/src/main/java/com/lagradost/cloudstream3/MainAPI.kt [hunk 2]
```diff
@@ -163,6 +163,38 @@ object APIHolder {
...
+     * Get anime tracker information based on title, year and type.
+     * Both titles are attempted to be matched with both Romaji and English title.
+     * Uses the consumet api.
+     *
+     * @param mainTitle Title to search by and match
+     * @param secondaryTitle Optional extra title if you have multiple titles and want extra guarantee to match.
+     * @param types Optional parameter to narrow down the scope to Movies, TV, etc. See TrackerType.getTypes()
+     * @param year Optional parameter to only get anime with a specific year
+     **/
+    suspend fun getTracker(
```
Comment (LagradOst) [inline app/src/main/java/com/lagradost/cloudstream3/MainAPI.kt:176]: make a cached version of this that stores them in a hashmap also make it safe via try
var trackerCache : HashMap<String, AniSearch> = hashMapOf()
suspend fun getTracker(...) : Tracker? {
try {
let mainTitle = titles[0]
let search = trackerCache.get(mainTitle) :? app.get("https://api.consumet.org/meta/anilist/$mainTitle")
            .parsedSafe<AniSearch>()?.also { trackerCache[mainTitle] = it } ?: return null
...
} catch(e : Throwable) {
return null
}
}

**Review by LagradOst — CHANGES_REQUESTED** [review]
I want this 
1. safe, the provider should still work if getTracker crashes
2. cached, no unnecessary network calls
3. expandable, provide a list of names

File: app/src/main/java/com/lagradost/cloudstream3/MainAPI.kt (hunk 1 above)
Comment (Blatzar) [reply app/src/main/java/com/lagradost/cloudstream3/MainAPI.kt:177]: True, true. I forgor :skull:

Comment (KillerDogeEmpire) [reply app/src/main/java/com/lagradost/cloudstream3/MainAPI.kt:177]: @Blatzar I understand how to do the other change, but don't understand what blob wants in this one. How would I make a list... I sound really dumb rn, I'm tired

**Pull request merged**